presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java

```diff
@@ -252,7 +260,53 @@ public static boolean canCoerce(Type actualType, Type expectedType)
     public static boolean canCoerce(TypeSignature actualType, TypeSignature expectedType)
     {
         Optional<TypeSignature> commonSuperTypeSignature = getCommonSuperTypeSignature(actualType, expectedType);
-        return commonSuperTypeSignature.isPresent() && commonSuperTypeSignature.get().equals(expectedType);
+        return commonSuperTypeSignature.isPresent() && typesMatch(commonSuperTypeSignature.get(), expectedType);
+    }
+
+    private static boolean typesMatch(TypeSignature actualType, TypeSignature expectedType)
```

Comment:
This code looks similar to `com.facebook.presto.metadata.Signature#matchAndBindParameters` and `com.facebook.presto.type.TypeUtils#resolveCalculatedType`. Maybe this should be unified somehow? Maybe one pass over Signature that will "in place"/"online"/"live" checking for matching, coercions and simultaneously calculating literal parameters?